src/assets/BrowserAutomationProtocol/browser-automation-protocol.md

```diff
@@ -0,0 +1,79 @@
+- **Data Extraction**: _Automated web scraping is a great and efficient way to extract data from websites._
+
+## Automating the browser
+
+The simplest way to automate any browser is to use the available automation libraries such as selenium, playwright, cypress, and many more. However, today we are not going to learn about those automation tools. Instead, we are here to learn about the different approaches we can use to directly control the browser without the use of any automation libraries.
+
+In order to automate the browser, there are mainly three protocols or let's say approaches in practice.
+
+1. WebDriver (Selenium WebDriver)
+2. Chrome DevTools Protocol
+3. Native Scripting (Embedded JavaScript)
+
+### WebDriver
+
+WebDriver is a remote control interface for controlling browsers, developed by Selenium HQ and later standardized by W3C. It provides a platform and language-neutral interface and supports almost all the major browsers such as Chrome, Firefox, Safari, Edge, Opera, etc.
+
+Communication with the browser is done through the JSON Wire protocol, a RESTful web service using JSON over HTTP. The server-side implementation of this protocol is known as the webdriver (aka browser driver); examples are chromedriver, geckodriver, etc. And the client-side implementation can be done in any preferred language and pattern. Some examples of webdriver clients are selenium, nightwatch, webdriverio, etc.
+
+![WebDriver Protocol](/src/assets/BrowserAutomationProtocol/images/webdriver_ill.jpg)
+
+Learn more about the protocol at [W3C WebDriver](https://w3c.github.io/webdriver/)
+
+### Chrome DevTools Protocol
+
+The Chrome DevTools Protocol (CDP) is a debugging protocol used by Chromium-based browsers such as Chrome, Edge, Opera, etc. Although it is not designed for testing purposes, it is a great tool for automating the browser and some automation tools such as puppeteer and playwright have proven that.
+
+There is no middleman like webdriver server (browser driver) between the client and the browser in this approach. Instead, the browser is directly controlled by the client using CDP. The communication with the browser is done through the socket connection and hence also enables bi-directional communication.
+
+![Chrome DevTools Protocol](/src/assets/BrowserAutomationProtocol/images/cdp_ill.jpg)
+
+Learn more about the protocol at [CDP](https://chromedevtools.github.io/devtools-protocol/)
+
+### Native Scripting
+
+This approach is different from above two protocols. In the above two approaches, the automation script (client) and the browser are separated and run in different processes. In contrast, in this approach, the automation script is embedded in the browser itself and hence the browser is directly controlled by the injected script.
+
+This comes with some trade-offs such as being unable to support multiple browser tabs and more than one browser cannot be controlled at a time. But it has its own uniqueness and strengths.
```

Comment:
what are those uniqueness and strengths?

Reply:
I will work for the next blog on this topic :+1: